Add order model

JohnGrey17 opened the pull request.

JohnGrey17 committed:
Created Order/OrderItem/Status models(raw)

JohnGrey17 committed:
Added changeLog with order/order_item table

JohnGrey17 committed:
implemented Order/OrderItem/Status/StatusConverter

JohnGrey17 committed:
Added RAW
*mapper
*Dto
*Repo
*Service
*Controller

for ORDER AND ORDER_ITEM

JohnGrey17 committed:
Added RAW
*mapper
*Dto
*Repo
*Service
*Controller

for ORDER AND ORDER_ITEM

JohnGrey17 committed:
Added RAW
*mapper
*Dto
*Repo
*Service
*Controller

for ORDER AND ORDER_ITEM

JohnGrey17 committed:
commit before merge with ShoppingCart
Branch

JohnGrey17 committed:
Merge branch 'Add-ShoppingCart-model' into add-order-model

JohnGrey17 committed:
Implemented POST API
(implemented method that implemented create new order)

JohnGrey17 committed:
Implemented all methods

JohnGrey17 committed:
Implemented all methods

andrii-hoienko commented on src/main/java/org/example/bookstore/controller/OrderController.java:
```diff
@@ -0,0 +1,59 @@
+    @PostMapping
+    @PreAuthorize("hasAnyRole('USER','ADMIN')")
+    public OrderResponseDto createNewOrder(Authentication authentication,
+                                           @RequestBody OrderRequestDto requestDto) {
```
```suggestion
                                           @RequestBody @Valid OrderRequestDto requestDto) {
```

andrii-hoienko replied:
Check other methods

andrii-hoienko commented on src/main/java/org/example/bookstore/controller/OrderController.java:
```diff
@@ -0,0 +1,59 @@
+        return orderService.getItemByOrderIdAndItemId(orderId, itemId);
+    }
+
+    @PutMapping("/{orderId}")
```
```suggestion
    @PatchMapping("/{orderId}")
```

andrii-hoienko reviewed (changes requested):
Please, CHECK PR BEFORE PUSHING to review, and prepare PR wiht necessary changes only

JohnGrey17 committed:
Added method into CartItemRepository that will be deleting all cartItems from the ShoppingCart after order create

JohnGrey17 committed:
Fixed

JohnGrey17 committed:
Fixed

yevheniisynytsia commented on src/main/java/org/example/bookstore/controller/OrderController.java:
```diff
@@ -0,0 +1,73 @@
+    @PreAuthorize("hasAnyRole('USER','ADMIN')")
+    @Operation(summary = "Get user's order history", description = "Retrieve the order history"
+            + " of the authenticated user.")
+    public List<OrderResponseDto> getUsersOrdersHistory(Pageable pageable) {
```
please check if user can get only his order history (add authentication)

yevheniisynytsia commented on src/main/java/org/example/bookstore/controller/OrderController.java:
```diff
@@ -0,0 +1,73 @@
+            + "the authenticated user.")
+    public OrderResponseDto createNewOrder(Authentication authentication,
+                                           @RequestBody @Valid OrderRequestDto requestDto) {
+        return orderService.createNewOrder(requestDto, authentication);
```
please pass userId instead of whole authentication

you may also take a look at `@AuthentiocationPrincipal`

yevheniisynytsia commented on src/main/java/org/example/bookstore/controller/OrderController.java:
```diff
@@ -0,0 +1,73 @@
+    @Operation(summary = "Change order status", description = "Change the status"
+            + " of an order by its ID.")
+    public OrderUpdatedDto changeStatusOfOrder(@PathVariable Long orderId,
+                                               @RequestBody OrderStatusUpdateRequest request) {
```
add valid

yevheniisynytsia commented on src/main/java/org/example/bookstore/dto/orderdto/OrderStatusUpdateRequest.java:
```diff
@@ -0,0 +1,8 @@
+
+@Data
+public class OrderStatusUpdateRequest {
+    private String status;
```
I think `@NotNull` should be here

yevheniisynytsia commented on src/main/java/org/example/bookstore/mapper/OrderMapper.java:
```diff
@@ -0,0 +1,22 @@
+import org.mapstruct.Mapping;
+
+@Mapper(config = MapperConfig.class, uses = {UserMapper.class,
+        BookMapper.class, OrderItemsMapper.class})
```
why do you need so many mappers here? I guess `OrderItemsMapper` is enough

yevheniisynytsia commented on src/main/java/org/example/bookstore/model/OrderItem.java:
```diff
@@ -0,0 +1,49 @@
+
+    @ManyToOne(fetch = FetchType.LAZY)
+    @JoinColumn(name = "book_id", nullable = false)
+    private Book book;
```
exclude from `toString`

you may also add equals and hashcode to all your entities and exclude lazy fields

yevheniisynytsia commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,150 @@
+        Set<CartItem> cartItems = shoppingCart.getCartItems();
+
+        if (cartItems == null || cartItems.isEmpty()) {
+            throw new EntityNotFoundException("Your shopping cart is empty ");
```
this is not truly `EntityNotFoundException`, let's create a specific exception

yevheniisynytsia commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,150 @@
+    }
+
+    @Override
+    @Transactional
```
please do not put `@Transactional` everywhere

let's think where it should be placed

andrii-hoienko replied:
Not resolved

yevheniisynytsia commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,150 @@
+        boolean statusFound = false;
+
+        for (Status sts : values) {
+            if (statusString.equalsIgnoreCase(sts.toString())) {
```
```suggestion
            if (statusString.equalsIgnoreCase(sts.name())) {
```

yevheniisynytsia commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,150 @@
+        Status[] values = Status.values();
+        boolean statusFound = false;
+
+        for (Status sts : values) {
```
```suggestion
        for (Status enumValue : values) {
```

yevheniisynytsia reviewed (changes requested):
Well done! Left comments to work on!

JohnGrey17 committed:
Added Docker

JohnGrey17 committed:
fixed solution
- fixed method getUsersOrdersHistory(added AuthenticationPrincipal)
- fixed createNewOrder method(start use AuthenticationPrincipal)
- add @Valid AND @NotBlank to changeStatusOfOrder
- removed unused mapper from OrderMapper
- Excluded ToString from OrderItem
- Added new Exception EmptyCartException to OrderServiceImpl and also added HANDLER t CustomGlobalExceptionHandler

JohnGrey17 committed:
fixed solution
- fixed method getUsersOrdersHistory(added AuthenticationPrincipal)
- fixed createNewOrder method(start use AuthenticationPrincipal)
- add @Valid AND @NotBlank to changeStatusOfOrder
- removed unused mapper from OrderMapper
- Excluded ToString from OrderItem
- Added new Exception EmptyCartException to OrderServiceImpl and also added HANDLER t CustomGlobalExceptionHandler

karma-o commented on .env:
env files should not be pushed to git, delete it from here and add it to gitignore

karma-o replied:
still here, try again, you can copy it, then delete in IDE, then push, then add it back and add it to the gitignore, please check if it is present in the PR before merging it

karma-o commented on src/main/java/org/example/bookstore/model/order/Order.java:
```diff
@@ -0,0 +1,62 @@
+    @JoinColumn(name = "user_id", nullable = false)
+    private User user;
+
+    @Convert(converter = StatusConverter.class)
+    @Column(nullable = false)
+    private Status status;
```
why not just use `@Enumereted(STRING)`?

karma-o commented on src/main/resources/db/changelog/changes/15-create-order-table.yaml:
```diff
@@ -0,0 +1,50 @@
+                  constraints:
+                    nullable: false
+              - column:
+                  name: status
+                  type: varchar(255)
```
you even have simple varchar type here, so that would be much simpler approach instead of using converters and stuff

karma-o commented on src/main/java/org/example/bookstore/model/status/Status.java:
```diff
@@ -0,0 +1,7 @@
+package org.example.bookstore.model.status;
+
+public enum Status {
```
```suggestion
public enum OrderStatus {
```
Or make it a nested class of `Order`, up to you

karma-o commented on src/main/java/org/example/bookstore/dto/orderdto/OrderUpdatedDto.java:
```diff
@@ -0,0 +1,5 @@
+package org.example.bookstore.dto.orderdto;
+
+public class OrderUpdatedDto {
+    private String status;
```
should work and validate without a problem, maybe add @JsonCreator with a `Status fromString(String str)` method to the enum class if it does not
```suggestion
    private Status status;
```

karma-o replied:
also add validation
```suggestion
    @NotNull
    private String status;
```

karma-o commented on src/main/java/org/example/bookstore/dto/orderdto/OrderUpdatedDto.java:
```diff
@@ -0,0 +1,5 @@
+package org.example.bookstore.dto.orderdto;
+
+public class OrderUpdatedDto {
```
```suggestion
public class OrderUpdateDto {
```

karma-o commented on src/main/java/org/example/bookstore/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -45,6 +45,13 @@ protected ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundExc
         return new ResponseEntity<>(body, status);
     }
 
+    @ExceptionHandler(EmptyCartException.class)
+    protected ResponseEntity<Object> handleEmptyCartException(EmptyCartException ex) {
```
you have numerous identical bad request handlers, combine them into 1
```suggestion
    @ExceptionHandler({
        OtherBadRequestException.class,
        EmptyCartException.class})
    protected ResponseEntity<Object> handleBadRequestException(RuntimeException ex) {
        // common handling logic
    }
```

karma-o commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,153 @@
+
+        Order order = new Order();
+        order.setUser(user);
+        order.setOrderDate(LocalDateTime.now());
```
You can use `@CreationTimestamp` in the entity class instead of setting it manually here, or a `@PrePersist` method there

karma-o commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,153 @@
+        order.setStatus(Status.PENDING);
+        order.setShippingAddress(requestDto.getShippingAddress());
+
+        Set<OrderItem> orderItems = cartItems.stream().map(cartItem -> {
+            OrderItem orderItem = new OrderItem();
+            orderItem.setOrder(order);
+
+            orderItem.setBook(cartItem.getBook());
+            orderItem.setQuantity(cartItem.getQuantity());
+            orderItem.setPrice(cartItem.getBook().getPrice().multiply(
+                    BigDecimal.valueOf(cartItem.getQuantity())));
+
+            return orderItem;
+        }).collect(Collectors.toSet());
```
move this to some map method or maybe even create a mapper class for that

karma-o commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,153 @@
+            return orderItem;
+        }).collect(Collectors.toSet());
+        order.setOrderItems(orderItems);
+        order.setTotal(orderItems.stream()
+                .map(orderItem -> orderItem.getPrice().multiply(
+                        BigDecimal.valueOf(orderItem.getQuantity())))
+                .reduce(BigDecimal.ZERO, BigDecimal::add));
```
extract total price calculation to another method

karma-o commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,153 @@
+    @Transactional
+    public OrderResponseDto createNewOrder(OrderRequestDto requestDto,
+                                           Long userId) {
+        User user = userRepository.findById(userId)
+                .orElseThrow(() -> new EntityNotFoundException("User with id: " + userId
+                        + "does not exist"));
+        ShoppingCart shoppingCart = shoppingCartRepository
+                .findShoppingCartByUserId(user.getId()).orElseThrow(()
+                        -> new EntityNotFoundException("Shopping cart does not exist"));
```
for the future, would be nice to encapsulate that existence check into the entity respective service methods like `service.getById(id)` to not do it here

karma-o commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,153 @@
+    @Transactional
+    public List<OrderResponseDto> getAllUserOrders(Long userId,Pageable pageable) {
+        User user = userRepository.findById(userId).orElseThrow(() -> new EntityNotFoundException(
+                "can`t find user"));
```
add id to the message

karma-o commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,153 @@
+    }
+
+    @Override
+    @Transactional
```
following up with @yevheniisynytsia comment, we can remove `@transactional` from the get operations, or set `readOnly = true` for them

karma-o reviewed (changes requested):
Just a few more things that we should fix

JohnGrey17 committed:
fixed solution

JohnGrey17 commented:
i did not make a special mapper for the method createNewOrder because I used in all method same approach with setter. and for me it is looks better to read the code

andrii-hoienko reviewed (changes requested):
Please resolve mentor comments first

JohnGrey17 committed:
fixed solution

JohnGrey17 committed:
fixed solution

andrii-hoienko commented on .env:
```diff
@@ -0,0 +1,13 @@
+MYSQL_DATABASE=books_store
```
![Uploading image.png…]()

JohnGrey17 replied:
+

andrii-hoienko commented on src/main/java/org/example/bookstore/model/order/Order.java:
```diff
@@ -0,0 +1,60 @@
+    private User user;
+
+    @Column(nullable = false, columnDefinition = "varchar(255)")
+    private Status status;
```
![image](https://github.com/JohnGrey17/bookStore/assets/69054906/cdcbdbe8-e691-4f43-a5cf-8b6f50fb21d0)

JohnGrey17 replied:
+

andrii-hoienko commented on src/main/java/org/example/bookstore/dto/orderdto/OrderUpdatedDto.java:
```diff
@@ -0,0 +1,9 @@
+import jakarta.validation.constraints.NotNull;
+import org.example.bookstore.model.status.Status;
+
+public class OrderUpdatedDto {
```
![Uploading image.png…]()

JohnGrey17 replied:
+

andrii-hoienko commented on src/main/java/org/example/bookstore/exception/CustomGlobalExceptionHandler.java:
```diff
@@ -45,6 +45,13 @@ protected ResponseEntity<Object> handleEntityNotFoundException(EntityNotFoundExc
         return new ResponseEntity<>(body, status);
     }
 
+    @ExceptionHandler(EmptyCartException.class)
```
![image](https://github.com/JohnGrey17/bookStore/assets/69054906/faae5f23-9879-4b4b-b457-b6c3d16fccd3)

JohnGrey17 replied:
+

andrii-hoienko commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,155 @@
+        }).collect(Collectors.toSet());
+        order.setOrderItems(orderItems);
+        order.setTotal(calculateTotal(orderItems));
+
```
![image](https://github.com/JohnGrey17/bookStore/assets/69054906/0d363423-94c8-47e3-a14e-08e4e007c31f)

JohnGrey17 replied:
+

andrii-hoienko commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,155 @@
+                new EntityNotFoundException("Order with id: " + orderId + " does not exist"));
+
+        String statusString = request.getStatus().trim().toUpperCase();
+        System.out.println("Received status from request: " + statusString);
```
If you want to add logs please use logger

JohnGrey17 replied:
+

andrii-hoienko commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,155 @@
+    @Transactional
+    public OrderUpdatedDto changeStatusOfOrderById(Long orderId,
+                                                   OrderStatusUpdateRequestDto request) {
+        Order order = orderRepository.findById(orderId).orElseThrow(() ->
+                new EntityNotFoundException("Order with id: " + orderId + " does not exist"));
+
+        String statusString = request.getStatus().trim().toUpperCase();
+        System.out.println("Received status from request: " + statusString);
+
+        Status[] values = Status.values();
+        boolean statusFound = false;
+
+        for (Status enumValue : values) {
+            if (statusString.equalsIgnoreCase(enumValue.name())) {
+                order.setStatus(enumValue);
+                statusFound = true;
+                break;
+            }
+        }
+
+        if (!statusFound) {
+            throw new IllegalArgumentException("Invalid status: " + statusString);
+        }
+
+        Order savedOrder = orderRepository.save(order);
+        return orderMapper.toUpdateDto(savedOrder);
```
```suggestion
        Order order = orderRepository.findById(orderId).orElseThrow(() ->
                new EntityNotFoundException("Order with id: " + orderId + " does not exist"));
           order.set(request.getStatus())

        Order savedOrder = orderRepository.save(order);
        return orderMapper.toUpdateDto(savedOrder);
```

JohnGrey17 replied:
+

JohnGrey17 committed:
fixed solution(hope)

andrii-hoienko commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,155 @@
+    @Override
+    @Transactional(readOnly = true)
+    public List<OrderResponseDto> getAllUserOrders(Long userId,Pageable pageable) {
+        User user = userRepository.findById(userId).orElseThrow(() -> new EntityNotFoundException(
+                "can`t find user " + userId));
```
This logic should belong to user service, or you can accept user as parameter in method to avoid DB requiest

JohnGrey17 replied:
+

andrii-hoienko commented on src/main/java/org/example/bookstore/controller/OrderController.java:
```diff
@@ -0,0 +1,76 @@
+    @Operation(summary = "Get all items by order ID", description = "Retrieve all items associated "
+            + "with a specific order ID.")
+    public List<OrderItemResponseDto> getAllOrderItemByOrderId(@PathVariable Long orderId) {
+        return orderService.getAllItemsByOrderId(orderId);
```
Please make sure that one user can not access data of different user

JohnGrey17 replied:
ALL WORKS GOOD

andrii-hoienko commented on src/main/java/org/example/bookstore/controller/OrderController.java:
```diff
@@ -0,0 +1,76 @@
+    public OrderItemResponseDto getSpecifiedItemInfo(@PathVariable Long orderId,
+                                                     @PathVariable Long itemId) {
+        return orderService.getItemByOrderIdAndItemId(orderId, itemId);
+    }
```
Please make sure that one user can not access data of different user

andrii-hoienko commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,155 @@
+    @Override
+    @Transactional(readOnly = true)
+    public List<OrderItemResponseDto> getAllItemsByOrderId(Long orderId) {
+        Order order = orderRepository.findById(orderId)
+                .orElseThrow(() -> new EntityNotFoundException("can`t find "
+                        + "order by id: " + orderId));
```
Code duplication

JohnGrey17 replied:
+

andrii-hoienko commented on src/main/java/org/example/bookstore/dto/orderdto/OrderRequestDto.java:
```diff
@@ -0,0 +1,10 @@
+
+@Data
+public class OrderRequestDto {
+    @Length(min = 7,max = 255)
+    private String shippingAddress;
```
```suggestion
    @Length(min = 7,max = 255)
    @NotBlank
    private String shippingAddress;
```

JohnGrey17 replied:
+

andrii-hoienko commented on src/main/java/org/example/bookstore/dto/orderdto/OrderRequestDto.java:
```diff
@@ -0,0 +1,10 @@
+package org.example.bookstore.dto.orderdto;
+
+import lombok.Data;
+import org.hibernate.validator.constraints.Length;
```
Wrong package

JohnGrey17 replied:
-

andrii-hoienko reviewed (changes requested):
Please read comments cerafully and check PR before pushing

JohnGrey17 committed:
Merge branch 'doker' into add-order-model

# Conflicts:
#	.gitignore

JohnGrey17 committed:
fixed

karma-o commented on Dockerfile:
```diff
@@ -0,0 +1,18 @@
+# Builder stage
```
docker file is not part of this task, we should also remove it from this pr

karma-o commented on docker-compose.yml:
```diff
@@ -0,0 +1,36 @@
+version: '3.8'
```
same

karma-o commented on pom.xml:
```diff
@@ -17,6 +17,10 @@
     </properties>
     <dependencies>
+        <dependency>
+            <groupId>org.springframework.boot</groupId>
```
same

karma-o reviewed (approved):
Approved, to not torture you with any more rejects :sweat_smile:  but please fix the comments that I have left before merging, and verify if they are actually fixed by self-review of your PR

karma-o commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+                .filter(e -> Objects.equals(e.getId(), itemId))
+                .map(orderItemsMapper::toDto)
+                .findFirst()
+                .orElseThrow(() -> new EntityNotFoundException("Can't find item by id: " + itemId));
```
what if some user sends a request to get an order that does not belong to that user? I think he will have such an ability, and we should restrict him from that

karma-o commented on src/main/java/org/example/bookstore/service/order/OrderServiceImpl.java:
```diff
@@ -0,0 +1,137 @@
+        return orderItems.stream()
+                .map(orderItemsMapper::toDto)
+                .toList();
+    }
```
same here

JohnGrey17 committed:
fixed change Status method

JohnGrey17 committed:
Fixed Solution
